kvserver: leader lease can't be reacquired immediately after restart

iskettaneh opened the pull request:
This commit causes the node to wait for previous acquired leases to expire after a restart. This works for expiration and leader leases, but not epoch leases since liveness record might not be found in cache. This doesn't change the current behaviour for epoch-based leases. However, it fixes the problem for both expiration-based and leader leases.

This is important to avoid situations where we might accept a write operation that violates a previously served future read with the previous lease before the restart (We lose the timestamp cache after restarts).

Release note: None

References: #59874

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141637)
<!-- Reviewable:end -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-:-OJTjxfZ2zzVEpth_Fo5:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/client_lease_test.go` line 1905 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJTj3V-8kSmOtkiDUvc:-OJTj3V-8kSmOtkiDUvd:b-ik4qjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fa2910f1434650472641266e6efc59ef77ec8c/pkg/kv/kvserver/client_lease_test.go#L1905)):*
> ```Go
> // restart (We lose the timestamp cache after the restart).
> //
> // Leader leases avoids this situation because a replica won't be able to
> ```

nit: avoid
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1964 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJTjYTCDoPsclbBTYh7:-OJTjYTCDoPsclbBTYh8:b-smvczn) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fa2910f1434650472641266e6efc59ef77ec8c/pkg/kv/kvserver/client_lease_test.go#L1964)):*
> ```Go
> 	}
> 
> 	_, supportEndTime := tc.GetFirstStoreFromServer(t, 0).
> ```

Is it possible to compare the actual lease expiration rather than the store liveness expiration. Obviously we expect them to be the same, but for the same of keeping the test at the lease level.

Also (only if it's easy to do) maybe we can have an actual future-time read? 


<!-- Sent from Reviewable.io -->

iskettaneh committed:
kvserver: exp/leader leases can't be reacquired immediately after restart

This commit causes the node to wait for previous acquired leases to expire
after a restart. This works for expiration and leader leases, but not epoch
leases since liveness record might not be found in cache. This doesn't
change the current behaviour for epoch-based leases. However, it fixes the
problem for both expiration-based and leader leases.

This is important to avoid situations where we might accept a write
operation that violates a previously served future read with the previous
lease before the restart (We lose the timestamp cache after restarts).

Release note: None

References: #59874

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/client_lease_test.go` line 1964 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJTjYTCDoPsclbBTYh7:-OJTlW901OzHzW4HZSaC:b-4xhy75) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fa2910f1434650472641266e6efc59ef77ec8c/pkg/kv/kvserver/client_lease_test.go#L1964)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Is it possible to compare the actual lease expiration rather than the store liveness expiration. Obviously we expect them to be the same, but for the same of keeping the test at the lease level.

Also (only if it's easy to do) maybe we can have an actual future-time read?
</blockquote></details>

I am using the actual lease expiration time now


<!-- Sent from Reviewable.io -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> But probably have Arul take a look too.

> Reviewed 1 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-:-OJxef_T7fzpA_Ehp0qI:brpyd76)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/replica_init.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxeCq2AdYI7c9wrTcM:-OJxeCq2AdYI7c9wrTcN:bz8wjo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/replica_init.go#L514)):*
> ```Go
> // minLeaseProposedTS past the previous lease expiration time. It does so by
> // sleeping until Clock().Now() is past the previous lease expiration time.
> // This works for expiration-based leases, and leader-leases but not for
> ```

It could still work for epoch leases if the cache is up-to-date, right? If so, maybe we can say for epoch leases it's best effort?
___
*[`pkg/kv/kvserver/replica_init.go` line 523 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxdyQQ6WF5ucm2xdjW:-OJxdyQREGOf2Qy2UBr9:bj0x5eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/replica_init.go#L523)):*
> ```Go
> 		st := r.leaseStatusAtRLocked(r.AnnotateCtx(context.TODO()), r.Clock().NowAsClockTimestamp())
> 		if st.OwnedBy(store.StoreID()) {
> 			// Only sleep if we were the previous lease owner.
> ```

Should we expand this comment to explain why we need the sleep and we can't just forward `minLeaseProposedTS`? 
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1911 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxeQgF2Zjqa3emX1ij:-OJxeQgF2Zjqa3emX1ik:b-76207j) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/client_lease_test.go#L1911)):*
> ```Go
> 		func(t *testing.T, leaseType roachpb.LeaseType) {
> 			if leaseType == roachpb.LeaseEpoch {
> 				// Doesn't work for epoch leases since the liveness record might not be
> ```

Presumably it's not easy to control what's in the cache for this test? If it's easy to do, we can probably show that this works for epoch leases in the limited case where the cache is up-to-date.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/replica_init.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxeCq2AdYI7c9wrTcM:-OK7Zs4B5R7HcKXfc2gr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/replica_init.go#L514)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

It could still work for epoch leases if the cache is up-to-date, right? If so, maybe we can say for epoch leases it's best effort?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_init.go` line 523 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxdyQQ6WF5ucm2xdjW:-OK7_BDk8cdjfRQkd7Xz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/replica_init.go#L523)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Should we expand this comment to explain why we need the sleep and we can't just forward `minLeaseProposedTS`?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1911 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OJxeQgF2Zjqa3emX1ij:-OK7_BiT9B1jIqM3Evhw:b-kvq98s) ([raw file](https://github.com/cockroachdb/cockroach/blob/de1a5cda7fd58527d05faf469a4e009b6328e1d9/pkg/kv/kvserver/client_lease_test.go#L1911)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Presumably it's not easy to control what's in the cache for this test? If it's easy to do, we can probably show that this works for epoch leases in the limited case where the cache is up-to-date.
</blockquote></details>

I didn't find an easy way to do this. I didn't want to spend a lot of time on it since the issue already exists with epoch-based leases today.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Mostly looks good. Nice work here! 

> Reviewed 1 of 2 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-:-OKD_cVY02-Tg7mKQQXU:b5m1669)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miraradeva)

___
*[`-- commits` line 5 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKDWy8tAPIQBT7OXB4D:-OKDWy8tAPIQBT7OXB4E:bz5pq0c):*
nit: why this doesn't work for epoch based leases could deserve more words. 
___
*[`pkg/kv/kvserver/replica_init.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKDZCvp7Evb3yvSqoiL:-OKDZCvp7Evb3yvSqoiM:b-fptwpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56916aae99e63bb1d094eecab3493a5209ce321/pkg/kv/kvserver/replica_init.go#L99)):*
> ```Go
> 		// lease that contradicts a future read served by the old lease before the
> 		// restart (we would have lost that timestamp cache).
> 		if err := r.forwardMinLeaseProposedTSPastLeaseExpirationMuLocked(store); err != nil {
> ```

Is there a reason we need to assign `r.mu.mninLeaseProposedTS` in `initRaftMuLockedReplicaMuLocked` first, and then sleep and forward it here? Can we just omit any forwarding by sleeping before we call `r.initRaftMuLockReplicaMuLocked`?
___
*[`pkg/kv/kvserver/replica_init.go` line 527 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKD_Y9J12nfINLezsgg:-OKD_Y9J12nfINLezsgh:b5cy1zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56916aae99e63bb1d094eecab3493a5209ce321/pkg/kv/kvserver/replica_init.go#L527)):*
> ```Go
> 			// Note that we need to sleep (instead of just forwarding the
> 			// minLeaseProposedTS) because we will run into assertions where the
> 			// lease proposed time is in the future compared to r.Clock().Now().
> ```

nit: let's add "we don't allow acquiring a lease that starts in the future" at the end here.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/replica_init.go` line 99 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKDZCvp7Evb3yvSqoiL:-OKDmyVTCy0ouULSq4xu:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56916aae99e63bb1d094eecab3493a5209ce321/pkg/kv/kvserver/replica_init.go#L99)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is there a reason we need to assign `r.mu.mninLeaseProposedTS` in `initRaftMuLockedReplicaMuLocked` first, and then sleep and forward it here? Can we just omit any forwarding by sleeping before we call `r.initRaftMuLockReplicaMuLocked`?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-:-OKRm9iH71ol88g9W5Ox:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @miraradeva)

___
*[`pkg/kv/kvserver/replica_init.go` line 348 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKRlQEP8FklvnE3lqEa:-OKRlQEP8FklvnE3lqEb:b-gciuo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87f02b9c16d6740863a864f9dace654188c491d/pkg/kv/kvserver/replica_init.go#L348)):*
> ```Go
> 			// node was restarted, we don't want to reacquire the lease with a start
> 			// time that overlaps the previous lease.
> 			// This is important so that we don't serve a write request with the new
> ```

nit: the double "this is important" reads a bit weird. Maybe change the second one to "This ensures".

Separately, no new line here.
___
*[`pkg/kv/kvserver/replica_init.go` line 518 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/141637#-OKRlucR6IHQe0CdFros:-OKRlucR6IHQe0CdFrot:b-2fkxsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c87f02b9c16d6740863a864f9dace654188c491d/pkg/kv/kvserver/replica_init.go#L518)):*
> ```Go
> 
> // waitForPreviousLeaseToExpire waits for the previous lease to expire. It does
> // so by sleeping until Clock().Now() is past the previous lease expiration.
> ```

nit: should this say `Clock.Now()` is in the future of the previous lease expiration?


<!-- Sent from Reviewable.io -->

iskettaneh commented:
TFTR!

bors r+
